internal/vault/vault.go

```diff
@@ -129,6 +129,54 @@ var (
 	errSealed = errors.New("vault: key store is sealed")
 )
 
+// Status returns the current state of the Hashicorp Vault instance.
+// In particular, whether it is reachable and the network latency.
+func (s *KeyStore) Status(ctx context.Context) (key.StoreState, error) {
+	state, err := key.DialStore(ctx, s.config.Endpoint)
+	if err != nil {
+		return key.StoreState{}, err
+	}
+	if state.State == key.StoreUnreachable {
+		return state, nil
+	}
+
+	// Vault is reachable over the network. Now, we fetch Vault health
+	// information to check whether it can serve requests.
+	// We use a custom version of the Client.Sys().Health() SDK function
+	// since we cannot pass our context.
+
+	var req = s.client.NewRequest(http.MethodGet, "/v1/sys/health")
+	// If the code is 400 or above it will automatically turn into an error,
+	// but the sys/health API defaults to returning 5xx when not sealed or
+	// initialized, so we force this code to be something else so we parse correctly
+	req.Params.Add("uninitcode", "299")
+	req.Params.Add("sealedcode", "299")
+	req.Params.Add("standbycode", "299")
+	req.Params.Add("drsecondarycode", "299")
+	req.Params.Add("performancestandbycode", "299")
+
+	resp, err := s.client.RawRequestWithContext(ctx, req)
+	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
```

Comment:
Should state be modified if we cannot make a health check?

Reply:
In this case `state.State == StoreReachable` (since we were able to dial Vault but failed to check Vault's health).
This seems to be appropriate to describe the current situation. Vault is reachable but (probably not available).